lib/driver.coffee

```diff
@@ -18,63 +18,87 @@ partitioninfo = require('partitioninfo')
 Promise = require('bluebird')
 fs = Promise.promisifyAll(require('fs'))
 fatfs = require('fatfs')
+ext2fs = Promise.promisifyAll(require('ext2fs'))
+magic = Promise.promisifyAll(require('mmmagic'))
 
 SECTOR_SIZE = 512
 
-###*
-# @summary Get a fatfs driver given a file descriptor
-# @protected
-# @function
-#
-# @param {filedisk.Disk} disk - filedisk.Disk instance
-# @param {Number} offset - offset of the image
-# @param {Number} size - size of the image
-# @returns {Object} the fatfs driver
-#
-# @example
-# disk = new filedisk.FileDisk(fd)
-# fatDriver = getDriver(disk, 0, 2048)
-###
-getDriver = (disk, offset, size) ->
-	sectorPosition = (sector) -> offset + sector * SECTOR_SIZE
-	{
-		sectorSize: SECTOR_SIZE
-		numSectors: size / SECTOR_SIZE
-		readSectors: (sector, dest, callback) ->
-			disk.read(dest, 0, dest.length, sectorPosition(sector), callback)
-		writeSectors: (sector, data, callback) ->
-			disk.write(data, 0, data.length, sectorPosition(sector), callback)
-	}
+MAGIC_OPTS = (
+	magic.MAGIC_NO_CHECK_TAR |
+	magic.MAGIC_NO_CHECK_ENCODING |
+	magic.MAGIC_NO_CHECK_TOKENS |
+	magic.MAGIC_NO_CHECK_CDF |
+	magic.MAGIC_NO_CHECK_TEXT |
+	magic.MAGIC_NO_CHECK_ELF |
+	magic.MAGIC_NO_CHECK_APPTYPE
+)
+DETECTOR = new magic.Magic(MAGIC_OPTS)
+
+KNOWN_FILESYSTEMS =
+	'node-ext2fs': [ 'ext2', 'ext3', 'ext4' ]
+	'fatfs': [ 'FAT' ]
+
+detectDriver = (mime) ->
+	for own driver, filesystems of KNOWN_FILESYSTEMS
+		for filesystem in filesystems
+			if mime.indexOf(filesystem) != -1
+				return driver
+
+detectFS = (disk, offset, size) ->
+	disk = Promise.promisifyAll(disk, multiArgs: true)
+	size = Math.min(size, 2048)
+	buf = Buffer.allocUnsafe(size)
```

Comment:
Is there a specific reason for using `.allocUnsafe()` here? This could trip up the detector if `bytesRead < size`.